### PR TITLE
dev/sg: add linter for preventing certain dependencies from being upgraded

### PR DESCRIPTION
**Pull request by bobheadxi**
Prevents issues like https://github.com/sourcegraph/zoekt/pull/330#issuecomment-1116857568

## Test plan

image

### CODE REVIEW

File: dev/sg/lints.go [hunk 1]
```diff
@@ -77,6 +80,72 @@ var allLintTargets = lintTargets{
 	},
 }
 
+func goModGuards() lint.Runner {
+	const header = "go.mod version guards"
+
+	var maxVersions = map[string]*semver.Version{
+		"github.com/prometheus/common": semver.MustParse("v0.32.1"),
```
Comment (daxmc99) [inline dev/sg/lints.go:87]: This feels like it should be separate since these are config values.

File: dev/sg/lints.go [hunk 2]
```diff
@@ -77,6 +80,72 @@ var allLintTargets = lintTargets{
 	},
 }
 
+func goModGuards() lint.Runner {
+	const header = "go.mod version guards"
+
+	var maxVersions = map[string]*semver.Version{
```
Comment (daxmc99) [inline dev/sg/lints.go:86]: What should we do when we don't have any `maxVersion` constraints? To disable this linter I think you would need to comment out this entire function or do a larger code change than just a single line.

File: dev/sg/lints.go (hunk 1 above)
Comment (bobheadxi) [reply dev/sg/lints.go:87]: Linter configuration should just live with the linter IMO, especially since it is custom

File: dev/sg/lints.go (hunk 2 above)
Comment (bobheadxi) [reply dev/sg/lints.go:86]: The next line covers this already! https://github.com/sourcegraph/sourcegraph/pull/34929/files#diff-55084b030df9cbb2edca08a6c8997d8342126cc1c40d4f5736971eac73cd95f9R90-R94

```go
	if len(maxVersions) == 0 {
		return func(ctx context.Context, s *repo.State) *lint.Report {
			return &lint.Report{Header: header, Output: "No guards currently defined"}
		}
	}
```

**Review by jhchabran — APPROVED** [review]
🎉

**Pull request merged**